outputpath = "Results/LibrarySize/BioTipV3/"

# This code performs CTS analysis on the expressional data of scRNA-seq, using BioTIP (v. 1.3.0).
# The GEO-downloaded (GSE130146) values were normalized by library size factors.
# This code drops more clusters, only focusing on the path towards smooth muscle and hemangiogenic
# R 3.12
# Generated by:  AJ Kinstlick (akinstlick@uchicago.edu)
# Last update: 08/24/2020



library(igraph)
library(BioTIP)
library(scater)
packageVersion("BioTIP")  #[1] '1.3.0'


setwd("D:/SingleCellAnalysis/GSE130146/")
#source("../../../source/BioTIP_update_3.3_02282020.R")


###############################################################
#### I already have all info stored in object dimred.sce
#### This is generated from libsf_later_analysis.R (finished in line 148)
load(file=paste0(outputpath,"beforechoosingnormalization.RData"))
###############################################################
#these clusters are dropped for libsf
#dropping endoderm
dropcols = colData(dimred.sce)$label == 10 |colData(dimred.sce)$label == 12 |
  colData(dimred.sce)$label == 13 | colData(dimred.sce)$label == 1 |
  colData(dimred.sce)$label == 11 | colData(dimred.sce)$label == 8

dimred.sce = dimred.sce[,!dropcols]

dim(dimred.sce)
#[1] 33456  1257

save(dimred.sce, file=paste0(outputpath,"dimred_sce_more_dropped_cols.RData"), compress=TRUE)
load(file=paste0("Results/LibrarySize/BioTipV2/optimized_test_sd_selection_2.RData"))


samplesL <- split(rownames(colData(dimred.sce)),f = colData(dimred.sce)$label)
if(any(tmp<10))  samplesL <- samplesL[-which(tmp<10)]
unlist((tmp <- lapply(samplesL, length)))
#   2   3   4   5   6   7   9
# 309  80 119 144  38 234 333 




####### 2) find biomarker ######################
# Pre-selection Transcript
cut.preselect = 0.1
cut.fdr = 0.2
cut.minsize = 40

#libsf.var is the top 4000 HVGs per cell from my dataset
#generated in line 88 of libsf_later_analysis.R
dat <- dimred.sce[libsf.var,]

#logcounts normalized by library size
logmat <- as.matrix(logcounts(dat))
dim(logmat)
# [1] 4000 1257

#this is without optimizing sd selection
set.seed(100101101)
test2 <- sd_selection(logmat,samplesL,cutoff=.05)
igraphL <- getNetwork(test2, fdr = cut.fdr)
# 1:199 nodes
# 2:196 nodes
# 3:192 nodes
# 4:196 nodes
# 5:198 nodes
# 6:188 nodes
# 7:198 nodes
# 8:199 nodes
# 9:195 nodes
# 10:182 nodes
# 11:167 nodes
# 12:128 nodes
# 13:196 nodes
cluster <- getCluster_methods(igraphL)
names(cluster)
membersL <- getMCI(cluster,test2, adjust.size = FALSE, fun='BioTIP')
names(membersL)
pdf(file=paste0(outputpath,"MCIPlot_NonOptimized_", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"),
    width=20, height=7)
plotBar_MCI(membersL, ylim=c(0,3), minsize = cut.minsize)
#dev.copy2pdf(file=paste0(outputpath,"FirstMCIBarPlot", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))
dev.off()


#this optimizes sd selection
set.seed(100101101)
test <- optimize.sd_selection(logmat, samplesL, cutoff=cut.preselect,
                              times=.75, percent=0.9)
save(test, file=paste0(outputpath,"optimized_test_sd_selection_2.RData"), compress=TRUE)
load(file=paste0(outputpath,"optimized_test_sd_selection_2.RData"))
names(test)
# [1] "2" "3" "4" "5" "6" "7" "9"
head(test[["2"]])


# Network Partition
igraphL <- getNetwork(test, fdr = cut.fdr)
# want at least 500 nodes, otherwise change fdr cutoff
# 2:285 nodes
# 3:264 nodes
# 4:281 nodes
# 5:275 nodes
# 6:260 nodes
# 7:268 nodes
# 9:286 nodes

cluster <- getCluster_methods(igraphL)
names(cluster)
# [1] "2" "3" "4" "5" "6" "7" "9"


# 2.1) Identifying CTS, new MCI score #########

membersL <- getMCI(cluster,test, adjust.size = FALSE, fun='BioTIP')
names(membersL)
#[1] "members" "MCI"     "sd"      "PCC"     "PCCo"  
pdf(file=paste0(outputpath,"MCIBarPlotFixedClusters_3_", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"),
    width=40, height=7)
plotBar_MCI(membersL, ylim=c(0,10))
plotBar_MCI(membersL, ylim=c(0,6), minsize = 30)
plotBar_MCI(membersL, ylim=c(0,3), minsize = cut.minsize)
#dev.copy2pdf(file=paste0(outputpath,"FirstMCIBarPlot", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))
dev.off()
#the numbers in the parenthesis: they are total number of clusters, no control of the cell (sample) size per  cluster

#top targets: 8, 5, 7

# get the statistics using the old MCI system
source("RCode/BioTIP_update_3.4_07012020.R")
?getMaxStats

topMCI = getTopMCI(membersL[["members"]], membersL[["MCI"]], membersL[["MCI"]], min=cut.minsize, n=3)
names(topMCI)

# get the state ID and MCI statistics for the two leading MCI scores per state
maxMCIms <- getMaxMCImember(membersL[["members"]],
                            membersL[["MCI"]],min =cut.minsize, n=3)
names(maxMCIms)
#[1] "idx"             "members"         "2topest.members" "3topest.members"

maxMCI = getMaxStats(membersL[['MCI']], maxMCIms[['idx']])
unlist(maxMCI)
#        2         3         4         5         9 
#1.6733620 0.7476863 0.4491501 1.9222827 1.7816465  

# 2         3         4         5         7         9 
# 1.6733620 0.7476863 0.4949974 1.9222827 2.0271806 1.7816465 

names(maxMCIms[["members"]][names(topMCI)])
# [1] "7" "5" "9"

CTS.Lib = getCTS(maxMCI[names(topMCI)], maxMCIms[["members"]][names(topMCI)])
# Length: 56
# Length: 54
# Length: 94

# Length: 48
# Length: 56
# Length: 54

tmp <- unlist(lapply(maxMCIms[['idx']][names(topMCI)], length))
#tmp calculates the number of bars within each named state
whoistop2nd <- names(tmp[tmp==2])
#this returns all the groups with exactly 2 bars

CTS.Lib = append(CTS.Lib, maxMCIms[["2topest.members"]][whoistop2nd])
names(CTS.Lib)
# [1] "5" "9" "2"

# [1] "7" "5" "9"

maxMCI <- maxMCI[names(CTS.Lib)[1:3]]
maxMCI
#        5        9        2 
# 1.922283 1.781646 1.673362

maxMCI <- c(maxMCI, getNextMaxStats(membersL[['MCI']], idL=maxMCIms[['idx']], whoistop2nd))
names(maxMCI)
# [1] "8" "5" "7" "5" "7"
names(CTS.Lib)
# [1] "8" "5" "7" "5" "7"

save(CTS.Lib, maxMCI, file=paste0(outputpath,"CTS_Lib_3.RData"))
load(file=paste0(outputpath,"CTS_Lib_2.RData"))


## estimate significance NOT repeat (takes a while)
dim(logmat)
# [1] 4000 1257
# M is precalculated correlation matrix, will be reused in the downstream simulation analysis
M <- cor.shrink(logmat, Y = NULL, MARGIN = 1, shrink = TRUE)
dim(M) # 4000 4000
save(M, file=paste0(outputpath,"CTS_ShrinkM_2.RData"), compress=TRUE) # save the file as its calculation runs a while
load(file=paste0(outputpath,"CTS_ShrinkM_2.RData"))

C = 1000
simuMCI = list()
for (i in 1:length(CTS.Lib)){
  n <- length(CTS.Lib[[i]])
  simuMCI[[i]] <- simulationMCI(n, samplesL, logmat,  B=C, fun="BioTIP",M=M)
}
names(simuMCI) = names(CTS.Lib)

save(simuMCI, file=paste0(outputpath,"SimuMCI_3_",C,"CTS_AT1.CellType", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))

# load("../../../result/GSE52583/T1_branch.CellType/GSE52583_GenePermutation_1000CTS_AT1.CellType0.05_fdr0.1_minsize30.RData")

par(mfrow=c(1,3))
for (i in 1:length(simuMCI)){
  plot_MCI_Simulation(maxMCI[i], simuMCI[[i]], las=2,ylim=c(0,3.2),
                      main=paste(names(maxMCI)[i], n, "genes",
                                 "\n","vs. ",C, "times of gene-permutation"),
                      which2point=names(maxMCI)[i])
}
dev.copy2pdf(file=paste0(outputpath,"plot_MCI_Simulation_RandomGene_3.pdf"))

######## 3)  Finding Tipping Point and verify using IC score  #################

##################################################
######  BioTIP score, shuffling genes ##############
C= 1000
SimResults_s = list()
SimResults_g = list()
BioTIP_scores = list()
for(i in 1:length(CTS.Lib)){
  CTS <- CTS.Lib[[i]]
  n <- length(CTS)
  BioTIP_scores[[i]] <- getIc(logmat, samplesL, CTS, fun="BioTIP", shrink=TRUE)
  
  SimResults_g[[i]]  <- simulation_Ic(n, samplesL, logmat, B=C,
                                      fun="BioTIP", shrink=TRUE)
  #### BioTIP score,  shulffing labelling
  x <- 2
  # ns <- length(samplesL[[x]])  # of cells at the state of interest
  SimResults_s[[i]] <- matrix(nrow=length(samplesL), ncol=C)
  rownames(SimResults_s[[i]]) = names(samplesL)
  for(j in 1:length(samplesL)) {
    ns <- length(samplesL[[j]])  # of each state
    SimResults_s[[i]][j,] <- simulation_Ic_sample(logmat, ns, Ic=BioTIP_score[x],
                                                  genes=CTS, B=C,
                                                  fun="BioTIP", shrink=TRUE)
  }
}
names(SimResults_g) = names(CTS.Lib)
names(SimResults_s) = names(CTS.Lib)
names(BioTIP_scores) = names(CTS.Lib)
save( SimResults_g, SimResults_s, BioTIP_scores, file=paste0(outputpath,"LibSF_IC_sim_BioTIP_3.RData"), compress=TRUE)
# load(paste0(outputpath,"LibSF_IC_sim_BioTIP_2.RData") 


pdf(file=paste0(outputpath,"matplot_density_ic_simulation_3.pdf"),width=10, height=7)  
par(mfrow=c(2,4))
## global matplot 

for(i in 1:length(BioTIP_scores)){
  n = length(CTS.Lib[[i]])
  plot_Ic_Simulation(BioTIP_scores[[i]], SimResults_g[[i]], las = 2, ylab="BioTIP",
                     main= paste(names(BioTIP_scores)[i],"(",n," transcripts)"),
                     fun="matplot", which2point=names(BioTIP_scores)[i])
  interesting = which(names(samplesL) == names(BioTIP_scores[i]))
  p = length(which(SimResults_g[[i]][interesting,] >= BioTIP_scores[[i]][names(BioTIP_scores)[i]]))
  p = p/ncol(SimResults_g[[i]])
  #first p value calculated for exactly at tipping point
  p2 = length(which(SimResults_g[[i]] >= BioTIP_scores[[i]][names(BioTIP_scores)[i]]))
  p2 = p2/ncol(SimResults_g[[i]])
  p2 = p2/nrow(SimResults_g[[i]])
  #second p value calculated across all statuses
  ## #local Kernel Density Plot  
  d <- density(SimResults_g[[i]]) # returns the density data
  plot(d, xlim=range(c(BioTIP_scores[[i]],SimResults_g[[i]])),
       main=paste("pLocal=",p,", pGlobal=",p2)) # plots the results
  abline(v=BioTIP_scores[[i]][names(BioTIP_scores)[i]], col="green")
}
dev.off() 



#p values (delta scores p-value)
p_g = vector()
for (i in 1:length(BioTIP_scores)){
  p_g[i] = plot_SS_Simulation(BioTIP_scores[[i]], SimResults_g[[i]], 
                              main = paste("Delta BioTIP",n,"genes"), 
                              ylab=NULL)
}
p_g
# [1] 0.000 0.005 0.000

# [1] 0.000 0.000 0.005


p_s = vector()
for (i in 1:length(BioTIP_scores)){
  p_s[i] = plot_SS_Simulation(BioTIP_scores[[i]], SimResults_s[[i]], 
                              main = paste("Delta BioTIP",n,"genes"), 
                              ylab=NULL)
}
p_s
# [1] 0.000 0.003 0.000

# [1] 0.171 0.000 0.005


## supporting GitHub plots ###      
pdf(file=paste0(outputpath,"IC_SS_Simulations_SimresultsG_3.pdf"),width=10, height=7)  
par(mfrow=c(2,2))
for(i in 1:length(BioTIP_scores)){
  interesting = which(names(samplesL) == names(BioTIP_scores[i]))
  plot_Ic_Simulation(BioTIP_scores[[i]], SimResults_g[[i]], las = 2, ylab="BioTIP", ylim=c(0,0.12),
                     main=paste(names(CTS.Lib)[i],"_",length(CTS.Lib[[i]]), "genes", "\n","vs. ",C, "gene-permutations"),
                     fun="boxplot", which2point= interesting)
  plot_SS_Simulation(BioTIP_scores[[i]], SimResults_g[[i]], 
                     main = paste("Delta BioTIP",n,"genes"), 
                     ylab=NULL,
                     xlim=range(c(BioTIP_scores[[i]],SimResults_g[[i]])))
}
dev.off()

pdf(file=paste0(outputpath,"IC_SS_Simulations_SimresultsS_3.pdf"),width=10, height=7)  
par(mfrow=c(2,2))
for(i in 1:length(BioTIP_scores)){
  interesting = which(names(samplesL) == names(BioTIP_scores[i]))
  plot_Ic_Simulation(BioTIP_scores[[i]], SimResults_s[[i]], las = 2, ylab="BioTIP", ylim=c(0, .5),
                     main=paste(names(CTS.Lib)[i],"_",length(CTS.Lib[[i]]), "genes", "\n","vs. ",C, "sample-permutations"),
                     fun="matplot", which2point= interesting)
  plot_SS_Simulation(BioTIP_scores[[i]], SimResults_s[[i]], 
                     main = "Delta BioTIP, label shuffling", 
                     ylab=NULL,
                     xlim=range(c(BioTIP_scores[[i]],SimResults_s[[i]])))
}
dev.off()



