outputpath = "Results/LibrarySize/"

# This code performs CTS analysis on the expressional data of scRNA-seq, using BioTIP (v. 1.3.0).
# The GEO-downloaded (GSE130146) values were normalized by library size factors.
# R 3.12
# Generated by:  AJ Kinstlick (akinstlick@uchicago.edu)
# Last update: 08/24/2020



library(igraph)
library(BioTIP)
library(scater)
packageVersion("BioTIP")  #[1] '1.3.0'


setwd("D:/SingleCellAnalysis/GSE130146/")
#source("../../../source/BioTIP_update_3.3_02282020.R")


###############################################################
#### I already have all info stored in object dimred.sce
#### This is generated from libsf_later_analysis.R (finished in line 148)
load(file=paste0(outputpath,"beforechoosingnormalization.RData"))
###############################################################
#these clusters are dropped for libsf
#dropping endoderm
dropcols = colData(dimred.sce)$label == 10 |colData(dimred.sce)$label == 12 |
  colData(dimred.sce)$label == 13

dimred.sce = dimred.sce[,!dropcols]

dim(dimred.sce)
#[1] 33456  1591

save(dimred.sce, file=paste0(outputpath,"dimred_sce_without_endoderm.RData"), compress=TRUE)


samplesL <- split(rownames(colData(dimred.sce)),f = colData(dimred.sce)$label)
unlist((tmp <- lapply(samplesL, length)))
#  1   2   3   4   5   6   7   8   9  11 
#222 309  80 119 144  38 234  83 333  29 
if(any(tmp<10))  samplesL <- samplesL[-which(tmp<10)]



####### 2) find biomarker ######################
# Pre-selection Transcript
cut.preselect = 0.1
cut.fdr = 0.2
cut.minsize = 50

#libsf.var is the top 4000 HVGs per cell from my dataset
#generated in line 88 of libsf_later_analysis.R
dat <- dimred.sce[libsf.var,]

#logcounts normalized by library size
logmat <- as.matrix(logcounts(dat))
dim(logmat)
# [1] 4000 1591

#this is without optimizing sd selection
set.seed(100101101)
test2 <- sd_selection(logmat,samplesL,cutoff=.05)
igraphL <- getNetwork(test2, fdr = cut.fdr)
# 1:199 nodes
# 2:196 nodes
# 3:192 nodes
# 4:196 nodes
# 5:198 nodes
# 6:188 nodes
# 7:198 nodes
# 8:199 nodes
# 9:195 nodes
# 10:182 nodes
# 11:167 nodes
# 12:128 nodes
# 13:196 nodes
cluster <- getCluster_methods(igraphL)
names(cluster)
membersL <- getMCI(cluster,test2, adjust.size = FALSE, fun='BioTIP')
names(membersL)
pdf(file=paste0(outputpath,"MCIPlot_NonOptimized_", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"),
    width=20, height=7)
plotBar_MCI(membersL, ylim=c(0,3), minsize = cut.minsize)
#dev.copy2pdf(file=paste0(outputpath,"FirstMCIBarPlot", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))
dev.off()


#this optimizes sd selection
set.seed(100101101)
test <- optimize.sd_selection(logmat, samplesL, cutoff=cut.preselect,
                              times=.75, percent=0.9)
save(test, file=paste0(outputpath,"optimized_test_sd_selection.RData"), compress=TRUE)
load(file=paste0(outputpath,"optimized_test_sd_selection.RData"))
names(test)
# [1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "11"
head(test[["1"]])


# Network Partition
igraphL <- getNetwork(test, fdr = cut.fdr)
# want at least 500 nodes, otherwise change fdr cutoff
#  1:275 nodes
#  2:263 nodes
#  3:265 nodes
#  4:278 nodes
#  5:280 nodes
#  6:273 nodes
#  7:273 nodes
#  8:306 nodes
#  9:273 nodes
# 11:248 nodes

cluster <- getCluster_methods(igraphL)
names(cluster)
# [1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "11"


# 2.1) Identifying CTS, new MCI score #########

membersL <- getMCI(cluster,test, adjust.size = FALSE, fun='BioTIP')
names(membersL)
#[1] "members" "MCI"     "sd"      "PCC"     "PCCo"  
pdf(file=paste0(outputpath,"MCIBarPlotFixedClusters", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"),
    width=40, height=7)
plotBar_MCI(membersL, ylim=c(0,10))
plotBar_MCI(membersL, ylim=c(0,6), minsize = 30)
plotBar_MCI(membersL, ylim=c(0,3), minsize = cut.minsize)
#dev.copy2pdf(file=paste0(outputpath,"FirstMCIBarPlot", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".pdf"))
dev.off()
#the numbers in the parenthesis: they are total number of clusters, no control of the cell (sample) size per  cluster

#top targets: 8, 5, 7

# get the statistics using the old MCI system
source("RCode/BioTIP_update_3.4_07012020.R")
?getMaxStats

topMCI = getTopMCI(membersL[["members"]], membersL[["MCI"]], membersL[["MCI"]], min=cut.minsize, n=3)
names(topMCI)

# get the state ID and MCI statistics for the two leading MCI scores per state
maxMCIms <- getMaxMCImember(membersL[["members"]],
                            membersL[["MCI"]],min =cut.minsize, n=3)
names(maxMCIms)
#[1] "idx"             "members"         "2topest.members" "3topest.members"

maxMCI = getMaxStats(membersL[['MCI']], maxMCIms[['idx']])
unlist(maxMCI)
#        1         5         7         8         9 
#0.7461443 2.6491770 1.7776188 3.1814009 1.0266997  

names(maxMCIms[["members"]][names(topMCI)])
# [1] "8" "5" "7"

CTS.Lib = getCTS(maxMCI[names(topMCI)], maxMCIms[["members"]][names(topMCI)])
# Length: 64
# Length: 55
# Length: 55

tmp <- unlist(lapply(maxMCIms[['idx']][names(topMCI)], length))
#tmp calculates the number of bars within each named state
whoistop2nd <- names(tmp[tmp==2])
#this returns all the groups with exactly 2 bars

CTS.Lib = append(CTS.Lib, maxMCIms[["2topest.members"]][whoistop2nd])
names(CTS.Lib)
# [1] "8" "5" "7" "5" "7"

maxMCI <- maxMCI[names(CTS.Lib)[1:3]]
maxMCI
# 8        5        7 
# 3.181401 2.649177 1.777619

maxMCI <- c(maxMCI, getNextMaxStats(membersL[['MCI']], idL=maxMCIms[['idx']], whoistop2nd))
names(maxMCI)
# [1] "8" "5" "7" "5" "7"
names(CTS.Lib)
# [1] "8" "5" "7" "5" "7"

save(CTS.Lib, maxMCI, file=paste0(outputpath,"CTS_Lib.RData"))
load(file=paste0(outputpath,"CTS_Lib.RData"))


## estimate significance NOT repeat (takes a while)
dim(logmat)
# [1] 4000 1591
# M is precalculated correlation matrix, will be reused in the downstream simulation analysis
M <- cor.shrink(logmat, Y = NULL, MARGIN = 1, shrink = TRUE)
dim(M) # 4000 4000
save(M, file=paste0(outputpath,"CTS_ShrinkM.RData"), compress=TRUE) # save the file as its calculation runs a while
load(file=paste0(outputpath,"CTS_ShrinkM.RData"))

C = 1000
simuMCI = list()
for (i in 1:length(CTS.Lib)){
  n <- length(CTS.Lib[[i]])
  simuMCI[[i]] <- simulationMCI(n, samplesL, logmat,  B=C, fun="BioTIP",M=M)
}
names(simuMCI) = names(CTS.Lib)

save(simuMCI, file=paste0(outputpath,"SimuMCI_",C,"CTS_AT1.CellType", cut.preselect, "_fdr",cut.fdr,"_minsize",cut.minsize,".RData"))

# load("../../../result/GSE52583/T1_branch.CellType/GSE52583_GenePermutation_1000CTS_AT1.CellType0.05_fdr0.1_minsize30.RData")

par(mfrow=c(1,5))
for (i in 1:length(simuMCI)){
  plot_MCI_Simulation(maxMCI[i], simuMCI[[i]], las=2,ylim=c(0,3.2),
                    main=paste(names(maxMCI)[i], n, "genes",
                               "\n","vs. ",C, "times of gene-permutation"),
                    which2point=names(maxMCI)[i])
}
dev.copy2pdf(file=paste0(outputpath,"plot_MCI_Simulation_RandomGene.pdf"))

######## 3)  Finding Tipping Point and verify using IC score  #################

##################################################
######  BioTIP score, shuffling genes ##############
C= 1000
SimResults_s = list()
SimResults_g = list()
BioTIP_scores = list()
for(i in 1:length(CTS.Lib)){
    CTS <- CTS.Lib[[i]]
    n <- length(CTS)
    BioTIP_scores[[i]] <- getIc(logmat, samplesL, CTS, fun="BioTIP", shrink=TRUE)
  
  SimResults_g[[i]]  <- simulation_Ic(n, samplesL, logmat, B=C,
                                 fun="BioTIP", shrink=TRUE)
  #### BioTIP score,  shulffing labelling
  x <- 2
  # ns <- length(samplesL[[x]])  # of cells at the state of interest
  SimResults_s[[i]] <- matrix(nrow=length(samplesL), ncol=C)
  rownames(SimResults_s[[i]]) = names(samplesL)
  for(j in 1:length(samplesL)) {
    ns <- length(samplesL[[j]])  # of each state
    SimResults_s[[i]][j,] <- simulation_Ic_sample(logmat, ns, Ic=BioTIP_score[x],
                                             genes=CTS, B=C,
                                             fun="BioTIP", shrink=TRUE)
  }
}
names(SimResults_g) = names(CTS.Lib)
names(SimResults_s) = names(CTS.Lib)
names(BioTIP_scores) = names(CTS.Lib)
save( SimResults_g, SimResults_s, BioTIP_scores, file=paste0(outputpath,"LibSF_IC_sim_BioTIP.RData"), compress=TRUE)
# load(paste0(outputpath,"LibSF_IC_sim_BioTIP.RData") 


pdf(file=paste0(outputpath,"matplot_density_ic_simulation.pdf"),width=10, height=7)  
par(mfrow=c(2,4))
## global matplot 

for(i in 1:length(BioTIP_scores)){
  n = length(CTS.Lib[[i]])
  plot_Ic_Simulation(BioTIP_scores[[i]], SimResults_g[[i]], las = 2, ylab="BioTIP",
                     main= paste(names(BioTIP_scores)[i],"(",n," transcripts)"),
                     fun="matplot", which2point=names(BioTIP_scores)[i])
  interesting = which(names(samplesL) == names(BioTIP_scores[i]))
  p = length(which(SimResults_g[[i]][interesting,] >= BioTIP_scores[[i]][names(BioTIP_scores)[i]]))
  p = p/ncol(SimResults_g[[i]])
  #first p value calculated for exactly at tipping point
  p2 = length(which(SimResults_g[[i]] >= BioTIP_scores[[i]][names(BioTIP_scores)[i]]))
  p2 = p2/ncol(SimResults_g[[i]])
  p2 = p2/nrow(SimResults_g[[i]])
  #second p value calculated across all statuses
  ## #local Kernel Density Plot  
  d <- density(SimResults_g[[i]]) # returns the density data
  plot(d, xlim=range(c(BioTIP_scores[[i]],SimResults_g[[i]])),
       main=paste("pLocal=",p,", pGlobal=",p2)) # plots the results
  abline(v=BioTIP_scores[[i]][names(BioTIP_scores)[i]], col="green")
  }
dev.off() 



#p values (delta scores p-value)
p_g = vector()
for (i in 1:length(BioTIP_scores)){
  p_g[i] = plot_SS_Simulation(BioTIP_scores[[i]], SimResults_g[[i]], 
                              main = paste("Delta BioTIP",n,"genes"), 
                              ylab=NULL)
}
p_g
# [1] 0.000 0.000 0.000 0.006 0.837

p_s = vector()
for (i in 1:length(BioTIP_scores)){
  p_s[i] = plot_SS_Simulation(BioTIP_scores[[i]], SimResults_s[[i]], 
                              main = paste("Delta BioTIP",n,"genes"), 
                              ylab=NULL)
}
p_s
# [1] 0.003 0.005 0.237 0.000 0.880



## supporting GitHub plots ###      
pdf(file=paste0(outputpath,"IC_SS_Simulations_SimresultsG.pdf"),width=10, height=7)  
par(mfrow=c(2,2))
for(i in 1:length(BioTIP_scores)){
interesting = which(names(samplesL) == names(BioTIP_scores[i]))
plot_Ic_Simulation(BioTIP_scores[[i]], SimResults_g[[i]], las = 2, ylab="BioTIP", ylim=c(0,0.12),
                   main=paste(names(CTS.Lib)[i],"_",length(CTS.Lib[[i]]), "genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= interesting)
plot_SS_Simulation(BioTIP_scores[[i]], SimResults_g[[i]], 
                   main = paste("Delta BioTIP",n,"genes"), 
                   ylab=NULL,
                   xlim=range(c(BioTIP_scores[[i]],SimResults_g[[i]])))
}
dev.off()

pdf(file=paste0(outputpath,"IC_SS_Simulations_SimresultsS.pdf"),width=10, height=7)  
par(mfrow=c(2,2))
for(i in 1:length(BioTIP_scores)){
interesting = which(names(samplesL) == names(BioTIP_scores[i]))
plot_Ic_Simulation(BioTIP_scores[[i]], SimResults_s[[i]], las = 2, ylab="BioTIP", ylim=c(0, .5),
                   main=paste(names(CTS.Lib)[i],"_",length(CTS.Lib[[i]]), "genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= interesting)
plot_SS_Simulation(BioTIP_scores[[i]], SimResults_s[[i]], 
                   main = "Delta BioTIP, label shuffling", 
                   ylab=NULL,
                   xlim=range(c(BioTIP_scores[[i]],SimResults_s[[i]])))
}
dev.off()


dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/AT1_BioTIP_vsSimulation_",length(CTS),"CTS.pdf"))
wilcox.test(simuBioTIP_g[1,], simuBioTIP_g[2,]) # p-value < 2.2e-16
wilcox.test(simuBioTIP_g[3,], simuBioTIP_g[2,]) # p-value < 2.2e-16


##################################################
######  Ic score, of the same number of random genes ##############
C= 1000
n <- length(CTS)
Ic_score <- getIc(dat, samplesL, CTS, fun="cor")
Ic_score
#    14.5     16.5 18.5_AT1 
# 1.076046 5.775612 2.280395 
simuIc_g  <- simulation_Ic(n, samplesL, dat, B=C,
                           fun='cor')


#### Ic score for the identified CTS,  shulffing labelling
x <- 2
#  ns <- length(samplesL[[x]])  # of fixed number of cells
simuIc_s <- matrix(nrow=length(samplesL), ncol=C)
rownames(simuIc_s) = names(samplesL)
for(j in 1:length(samplesL)) { 
  ns <- length(samplesL[[j]])  # for each state rewpectively 
  simuIc_s[j,] <- simulation_Ic_sample(dat, ns, Ic=Ic_score[x],
                                       genes=CTS, B=C,
                                       fun="cor")
}
# save( simuIc_g,  simuIc_s, file="F:/projects/BioTIP/result/GSE52583/AT1_simuIC.RData", compress=TRUE)

# load(file="F:/projects/BioTIP/result/GSE52583/AT1_simuIC.RData")  


par(mfrow=c(2,2))

plot_Ic_Simulation(Ic_score, simuIc_g, las = 2, ylab="Ic",
                   main=paste(length(CTS),names(samplesL)[x], "genes", "\n","vs. ",C, "gene-permutations"),
                   fun="boxplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_g, 
                   main = paste("Delta Ic",n,"genes"), 
                   ylab=NULL)
plot_Ic_Simulation(Ic_score, simuIc_s, las = 2, ylab="Ic", ylim=c(1, 10),
                   main=paste(length(CTS), names(samplesL)[x],"genes", "\n","vs. ",C, "sample-permutations"),
                   fun="matplot", which2point= x)
plot_SS_Simulation(Ic_score, simuIc_s, 
                   main = "Delta BioTIP, label shuffling", 
                   ylab=NULL)

dev.copy2pdf(file=paste0("ROutputFigs/GSE52583/AT1_Ic_vsSimulation_",length(CTS),"CTS.pdf"))
wilcox.test(simuIc_g[1,], simuIc_g[2,]) # p-value < 2.2e-16
wilcox.test(simuIc_g[3,], simuIc_g[2,]) # p-value < 2.2e-16


#############################################
### 3) motif analysis , do not repeat ####################### 

# # https://www.bioconductor.org/packages/release/workflows/vignettes/generegulation/inst/doc/generegulation.html
# library(Biostrings)
# library(GenomicFeatures)
# 
# load(file="GSE52583_annot.cds_Clustered_thresholded.RData")
# dim(annot.cds) # [1] 22854   196
# 
# G <- fData(annot.cds)
# dim(G) #[1] 22854    11
# colnames(G)
# # [1] "ensembl_gene_id"     "gene_short_name"     "chromosome_name"     "strand"             
# # [5] "start_position"      "end_position"        "gene_biotype"        "GSE52583.gene.id"   
# # [9] "locus"               "num_cells_expressed" "use_for_ordering"   
# 
# annot.cds <-annot.cds[G$use_for_ordering,]
# 
# dat <- exprs(annot.cds)
# dim(dat) #dim(dat) # [1] 10359   196
# G <- subset(G, use_for_ordering==TRUE)
# dim(G)  #[1] 10359    11
# 
# 
# 
#   resPath = 'FASTA_TSS_1500_500'
#   upstream=1500
#   downstream=500
#   GS <- CTS.AT1.E16
# 
#   x <- intersect(GS, rownames(dat))
#   cat(paste(length(x), 'out of', length(GS),'genes of interest are expressed', '\t'))
#   chromosomal.loc <- GRanges(G[x,]$locus)
#   names(chromosomal.loc) <- rownames(G[x,])
#     
#     ## get promoter loci and write into bed file
#   promoters <- promoters(chromosomal.loc, upstream=upstream, downstream=downstream)
#   # remove blacklist
#   blacklist.mm10 <- import('F:\\projects\\Share\\Blacklisted\\2019\\mm10-blacklist.v2.bed')
#   blacklisted <- findOverlaps(promoters, blacklist.mm10, type="within")
#   cat(length(blacklisted),'in blacklist', '\n')
#   if(length(blacklisted)>0)     promoters <- promoters[-from(blacklisted)]
#   export.bed(promoters, con="outPut.Rdata/GSE52583/CTS_AT2.E18.5.bed")
#   # 71 out of 71 genes of interest are expressed 3 in blacklist   

# Homer findMotifs.pl  CTS_AT1.E16.5.bed mm10 GSE115575_1500_500/ 
